Adding aggmo optimizer

AtheMathmo opened the pull request:
Implementing Aggregated Momentum optimizer ( https://arxiv.org/abs/1804.00325 )

I had some issues setting up a dev environment to run the tests. I'm hoping it will pass the CI tests but if not I will spend some time figuring out what is going wrong on my end.

AtheMathmo committed:
Adding aggmo optimizer

AtheMathmo committed:
Fixing whitespace issues

AtheMathmo committed:
More whitespace fixes

ezyang commented:
```
14:51:23 ======================================================================
14:51:23 ERROR: test_aggmo (__main__.TestOptim)
14:51:23 ----------------------------------------------------------------------
14:51:23 Traceback (most recent call last):
14:51:23   File "test_optim.py", line 278, in test_aggmo
14:51:23     wrap_old_fn(old_optim.aggmo, learningRate=1e-3)
14:51:23   File "test_optim.py", line 60, in _test_rosenbrock
14:51:23     params_t, state)
14:51:23   File "test_optim.py", line 28, in wrapper
14:51:23     return old_fn(closure, params, config, state)
14:51:23   File "/opt/python/2.7/lib/python2.7/site-packages/torch/legacy/optim/aggmo.py", line 79, in aggmo
14:51:23     x.add_(-clr, dfdx)
14:51:23 TypeError: add_() received an invalid combination of arguments - got (float, int), but expected one of:
14:51:23  * (float other, float alpha)
14:51:23       didn't match because some of the arguments have invalid types: (float, int)
14:51:23  * (Tensor other, float alpha)
14:51:23       didn't match because some of the arguments have invalid types: (!float!, int)
```

AtheMathmo committed:
Fixing legacy test issues

AtheMathmo commented:
I managed to figure out my environment and fixed (almost) all of the tests.

One of the tests fails; when the params fed to the constructor is `None`. I can't figure out why these tests work for the other optimizers.

soumith commented:
@AtheMathmo We dont think this method should be in pytorch core, as opposed to your own personal repository.

Our reservation is that we want to include methods that the community uses as a standard, or else the code maintenance problem balloons up for us. We do show discretion based on what the paper shows as evidence, for example BatchNorm was included within weeks of it's publish (in Torch). In terms of rejected methods, we've rejected newly minted papers such as Swish ( https://github.com/pytorch/pytorch/pull/3260 , https://github.com/pytorch/pytorch/pull/3182 ).

Do you have a differing opinion? Do you think the paper shows evidence that makes it a method that has obvious long-term success?

AtheMathmo commented:
@soumith TL;DR - I would probably agree. It does not make sense to encourage everyone publishing a novel method to shove their code into pytorch.

----

I am first author on the paper but will try to produce a balanced response.

I believe that the paper provides strong evidence that this method can provide benefits over existing approaches, e.g., classical momentum. It also shows that this has it's place even compared to adaptive methods, e.g., adam. Of course, I would much rather people read the paper themselves and decide whether they agree with these statements. And to try and answer your final question - I think it is too hard to say. We found that Nesterov momentum works very well but yet it seems to see relatively little use in practice.

My motivation for creating the pull request (excluding the more obvious and self-serving) is to encourage reproducibility. Making the barrier-to-entry as low as possible is an obvious advantage for new science and pytorch seems like a good way to achieve this. However, I understand that this may not align with your goals in this case.

Finally, as an optimization tool this seems to be relatively standalone. I would imagine this makes maintenance a little easier but I am not a core pytorch dev and would certainly trust your experience.

jamestwebber commented:
Not a stake-holder on either side, but I'd like to 👍 the notion that a low barrier to entry is important. Case in point: I was interested in trying this, so I googled "aggregated momentum pytorch" and found this PR.

I'm unlikely to implement it myself as I don't have the coding chops or the time to tinker with PyTorch internals. (I feel the same about [HALP](https://arxiv.org/abs/1803.03383), another method that sounds great but is difficult to implement on my own)

Definitely understand the reluctance to add lots of new methods to PyTorch. Is there a framework for writing compatible extensions, such that @AtheMathmo could create a separate repo that drops in a new optimizer?

AtheMathmo commented:
@jamestwebber I can certainly create my own repo containing the optimizer code. I think it should work okay if this is a separate module which users are to install (but it might be tricky to match against the correct version of pytorch in the future). My only issue is that I couldn't figure out a nice way to integrate with the pytorch tests.

@soumith Have you made a decision based on my response above?

ishaqadenali commented on torch/legacy/optim/aggmo.py:
```diff
@@ -0,0 +1,84 @@
+            for b in mom:
+                state['dfdx'][b] = torch.zeros_like(dfdx)
+        for b in mom:
+            state['dfdx'][b].mul_(b).add_(dfdx)
```
Hi @AtheMathmo, Firstly, great paper, it was super fun to read! Somehow I ended up here and am skimming through this code... Wondering why you don't add -dfdx and then omit the negative sign on the clr variable below.

From what I see you're doing:
Vt = beta*Vt-1 + grad(f(x))
x = x - gamma*VT

AtheMathmo replied:
I did this only to be consistent with the code here: https://github.com/pytorch/pytorch/blob/master/torch/legacy/optim/sgd.py#L62-L84

I'm not sure if there is any deeper meaning to this!

jamestwebber commented on torch/optim/aggmo.py:
```diff
@@ -0,0 +1,70 @@
+        if not 0.0 <= lr:
+            raise ValueError("Invalid learning rate: {}".format(lr))
+        for i, mom in enumerate(momentum):
+            if not 0.0 <= momentum:
```
This should be `mom`, not momentum. Also `mom` is a weird name for a variable. :p

soumith commented:
cc: @vincentqb can you review this and get it to completion if it makes sense. Use the guidelines that I shared with you separately.

Now that sufficient time has passed in the paper being published, it's time to re-evaluate, considering whether the paper has become standardized across the field, or if it still belongs to contrib

AtheMathmo commented:
@soumith thank you for reviving this! I'm happy to update the PR to pass the tests and should have time to do this in the next couple of days.

Note that the corresponding paper was recently published in ICLR2019.

soumith commented:
@AtheMathmo i wouldn't want you to spend more time until we give you a clear answer on whether we will accept it into core or not. @vincentqb will make this determination, you can wait for his answer before reviving the tests etc.

dcasbol committed:
Fix bug in momentum check.

It was checked against the whole list instead of the element, which gave a runtime error. I also simplified the conditions for improved legibility.

dcasbol committed:
Allow momentum and weight_decay == 0

AtheMathmo committed:
Merge pull request #1 from dcasbol/patch-1

Fix bug in momentum check.

iramazanli commented:
Thank you for preparing this amazing PR !

Indeed AggMo is a very successful and promising method. However, currently we are adding Optimizers according to rulebook given in :  https://github.com/pytorch/pytorch/wiki/Developer-FAQ#i-have-implemented-an-algorithm--feature-from-a-research-paper-that-is-related-to-optimization-when-is-the-correct-time-to-add-it-to-pytorch-core-

and we wait Optimization methods to satisfy these conditions before adding them to PyTorch. In that respect, we currently decide to wait before adding this method to Core Library.
